### DISCUSSION TITLE
all optional fields

### DISCUSSION THREAD

**Discussion by draego** [protobuf-c/protobuf-c#662] [Q&A] [answered]
I'm writing test code for a message proto composed of all "optional" fields.
Everything builds OK, but when I try to ...__get_packed_size(&...); the length is always 0, even if I have already set values.

If I make the first field "required" then all is well. Is this by design?

Thanks,
Steven

**Comment by draego** [comment]
NOTE: we tried this same thing in the c++ version and it appears to work OK. Point being that it doesn't work in the C version.

**Comment by edmonds** [comment]
Hi, Steven:

Can you show your test .proto and the code you're using to set the fields?

Off hand I would guess you're not turning on the corresponding `.has_*` variables in the struct...

**Comment by draego** [comment]
Edmonds,

Thanks for your speedy response. I actually did this with your first "AMessage" example by just changing the first "a" parm to also be optional. I also put in a package, but that didn't change anything:

> > > > > > > > > amessage.proto >>>>>>>>>>>>>>>
> > > > > > > > > package telecore.pnvc.command;

message AMessage
{
  optional int32 a=1 [default = 10];
  optional int32 b=2 [default = 11];
}
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

> > > > > > > > > > > amessage.c >>>>>>>>>>>>>>>>>>
> > > > > > > > > > > #include <stdio.h>
> > > > > > > > > > > #include <stdlib.h>
> > > > > > > > > > > #include "amessage.pb-c.h"

int main (int argc, const char \* argv[]) 
{
    Telecore__Pnvc__Command__AMessage msg = TELECORE__PNVC__COMMAND__AMESSAGE__INIT; // AMessage
    void *buf;                     // Buffer to store serialized data
    unsigned len;                  // Length of serialized data
        
    if (argc != 2 && argc != 3)
    {   // Allow one or two integers
        fprintf(stderr,"usage: amessage a [b]\n");
        return 1;
    }
        
    msg.a = atoi(argv[1]);
    if (argc == 3) { msg.has_b = 1; msg.b = atoi(argv[2]); }
    len = telecore__pnvc__command__amessage__get_packed_size(&msg);
        
    buf = malloc(len);
    telecore__pnvc__command__amessage__pack(&msg,buf);
        
    fprintf(stderr,"Writing %d serialized bytes\n",len); // See the length of message
    fwrite(buf,len,1,stdout); // Write to stdout to allow direct command line piping

    int\* ip = buf;
    fprintf(stderr,"done (0x%x) \n", *ip); 
        
    free(buf); // Free the allocated serialized buffer
    return 0;
}

<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

Thanks,
Steven

**Comment by edmonds** [comment]
My reading of the [Optional Fields And Default Values](https://developers.google.com/protocol-buffers/docs/proto#optional) section in the language guide is that default values for optional fields are used when _parsing_ a serialized message and an optional field isn't present. So there's no need to write the default value when _generating_ the serialized message.

**Comment by draego** [comment]
Sorry, but the reference to default values confuses me. I added those to see if the problem was fixed and it wasn't - the problem exists with or without default values.

The problem is that the length value returned from ...__get_packed_size() is 0 when both fields are marked optional.

Thanks,
Steven

**Comment by edmonds** [comment, marked as answer]
That's what optional means. If a field is optional and unset, it's omitted from the packed message. If your message consists entirely of unset, optional fields, then the packed size is going to be zero.

If you want the optional `a` and `b` fields in your example to be serialized onto the wire, you have to do `msg.has_a = 1; msg.has_b = 1;`.

**Comment by draego** [comment]
Gotcha - thanks. It was just counter-intuitive to me to have to do both to make the value "stick". Still new and learning this so gotcha's like this are to be expected.

Thanks for your help,
Steven
